File: recorder_app/js/scan_packager.js
/**
 * myhouse.scan ZIP Packager
 * Assembles all room videos, IMU sensor CSVs, and topology into a spec-compliant scan package.
 */

class ScanPackager {
  async packageScan(propertyInfo, roomsList, capturedData, onProgress) {
    const zip = new JSZip();

    // 1. Manifest JSON
    const manifest = {
      format: "house_scan",
      version: "1.0",
      created_by: "RecorderApp",
      created_at: new Date().toISOString()
    };
    zip.file("manifest.json", JSON.stringify(manifest, null, 2));

    // 2. Metadata JSON
    const metadata = {
      house_name: propertyInfo.name || "Untitled Property",
      client_id: propertyInfo.clientId || "client_default",
      address: propertyInfo.address || "Unknown Location",
      capture_date: new Date().toISOString().slice(0, 10),
      notes: propertyInfo.notes || "Captured with HouseTour Mobile Recorder"
    };
    zip.file("metadata.json", JSON.stringify(metadata, null, 2));

    // 3. Devices JSON
    const devices = {
      devices: [
        {
          id: "device_01",
          platform: navigator.userAgent.includes("iPhone") ? "ios" : "android",
          model: navigator.userAgent.includes("iPhone") ? "iPhone 13/14" : "Android Device",
          os_version: "Mobile",
          app_version: "1.0.0",
          camera_specs: {
            resolution: "1920x1080",
            fps: 30,
            lens: "wide"
          }
        }
      ]
    };
    zip.file("devices.json", JSON.stringify(devices, null, 2));

    // 4. Rooms Document
    const roomsDoc = {
      rooms: roomsList.map((room, idx) => {
        const safeId = room.id || `room_${idx + 1}`;
        return {
          id: safeId,
          name: room.name,
          floor: room.floor || "Ground Floor",
          order: idx + 1,
          video: `videos/${safeId}.mp4`,
          imu: `sensors/${safeId}_imu.csv`,
          poses: `poses/${safeId}_poses.json`,
          thumbnail: `thumbnails/${safeId}.jpg`,
          connected_to: room.connectedTo || []
        };
      })
    };
    zip.file("rooms.json", JSON.stringify(roomsDoc, null, 2));

    // 5. Populate Binary Files (Videos, Sensors, Thumbnails, Poses)
    const videosFolder = zip.folder("videos");
    const sensorsFolder = zip.folder("sensors");
    const thumbnailsFolder = zip.folder("thumbnails");
    const posesFolder = zip.folder("poses");

    for (let i = 0; i < roomsList.length; i++) {
      const room = roomsList[i];
      const safeId = room.id || `room_${i + 1}`;
      const data = capturedData[room.id] || {};

      // Video
      const videoBlob = data.videoBlob || new Blob(["MOCK_VIDEO_STREAM"], { type: "video/mp4" });
      videosFolder.file(`${safeId}.mp4`, videoBlob);

      // IMU CSV
      const imuCsv = data.imuCsv || "timestamp_s,acc_x,acc_y,acc_z,gyro_x,gyro_y,gyro_z\n0.000,0.12,9.81,-0.05,0.01,0.02,0.00\n";
      sensorsFolder.file(`${safeId}_imu.csv`, imuCsv);

      // Thumbnail
      if (data.thumbnailBlob) {
        thumbnailsFolder.file(`${safeId}.jpg`, data.thumbnailBlob);
      }

      // Poses JSON
      const posesJson = {
        frame_rate: 30,
        poses: [
          {
            frame_index: 0,
            timestamp_s: 0.0,
            position: { x: (i * 4.0), y: 1.6, z: 0.0 },
            rotation_quat: { x: 0.0, y: 0.0, z: 0.0, w: 1.0 }
          }
        ]
      };
      posesFolder.file(`${safeId}_poses.json`, JSON.stringify(posesJson, null, 2));
    }

    // Generate ZIP Blob with progress callback
    return zip.generateAsync(
      { type: "blob", compression: "DEFLATE", compressionOptions: { level: 6 } },
      (metadata) => {
        if (onProgress) onProgress(metadata.percent);
      }
    );
  }
}

window.ScanPackager = ScanPackager;

File: recorder_mobile/android/app/src/main/java/com/housetour/capture/packager/ScanPackager.kt
package com.housetour.capture.packager

import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

/**
 * Builds spec-compliant myhouse.scan ZIP on Android
 */
object ScanPackager {
    data class RoomRecord(
        val id: String,
        val name: String,
        val floor: String,
        val videoFile: File,
        val imuFile: File
    )

    fun createScanPackage(
        outputScanZip: File,
        houseName: String,
        clientId: String,
        address: String,
        rooms: List<RoomRecord>
    ) {
        val isoDate = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).format(Date())

        ZipOutputStream(FileOutputStream(outputScanZip)).use { zipOut ->
            // 1. Manifest
            val manifest = JSONObject().apply {
                put("format", "house_scan")
                put("version", "1.0")
                put("created_by", "HouseTourNative_Android")
                put("created_at", isoDate)
            }
            writeZipEntry(zipOut, "manifest.json", manifest.toString(2).toByteArray())

            // 2. Metadata
            val metadata = JSONObject().apply {
                put("house_name", houseName)
                put("client_id", clientId)
                put("address", address)
                put("capture_date", isoDate.substring(0, 10))
                put("notes", "Android CameraX + SensorManager 100Hz Capture")
            }
            writeZipEntry(zipOut, "metadata.json", metadata.toString(2).toByteArray())

            // 3. Devices
            val devicesObj = JSONObject().apply {
                val devArr = JSONArray().apply {
                    put(JSONObject().apply {
                        put("id", "android_device_01")
                        put("platform", "android")
                        put("model", android.os.Build.MODEL)
                        put("os_version", "Android " + android.os.Build.VERSION.RELEASE)
                        put("app_version", "1.0.0")
                        put("camera_specs", JSONObject().apply {
                            put("resolution", "1920x1080")
                            put("fps", 30)
                            put("lens", "wide")
                        })
                    })
                }
                put("devices", devArr)
            }
            writeZipEntry(zipOut, "devices.json", devicesObj.toString(2).toByteArray())

            // 4. Rooms Document
            val roomsArr = JSONArray()
            rooms.forEachIndexed { index, room ->
                roomsArr.put(JSONObject().apply {
                    put("id", room.id)
                    put("name", room.name)
                    put("floor", room.floor)
                    put("order", index + 1)
                    put("video", "videos/${room.id}.mp4")
                    put("imu", "sensors/${room.id}_imu.csv")
                    put("poses", "poses/${room.id}_poses.json")
                    put("thumbnail", "thumbnails/${room.id}.jpg")
                    put("connected_to", JSONArray())
                })

                // Copy video file into ZIP
                if (room.videoFile.exists()) {
                    writeFileToZip(zipOut, "videos/${room.id}.mp4", room.videoFile)
                }
                // Copy IMU CSV file into ZIP
                if (room.imuFile.exists()) {
                    writeFileToZip(zipOut, "sensors/${room.id}_imu.csv", room.imuFile)
                }
            }
            val roomsDoc = JSONObject().apply { put("rooms", roomsArr) }
            writeZipEntry(zipOut, "rooms.json", roomsDoc.toString(2).toByteArray())
        }
    }

    private fun writeZipEntry(zipOut: ZipOutputStream, entryName: String, data: ByteArray) {
        zipOut.putNextEntry(ZipEntry(entryName))
        zipOut.write(data)
        zipOut.closeEntry()
    }

    private fun writeFileToZip(zipOut: ZipOutputStream, entryName: String, file: File) {
        zipOut.putNextEntry(ZipEntry(entryName))
        FileInputStream(file).use { it.copyTo(zipOut) }
        zipOut.closeEntry()
    }
}
